### PR TITLE
Allow using custom configuration sources in sdk-autoconfigure

### PR DESCRIPTION
**Pull request by mateuszrzeszutek**
Fixes https://github.com/open-telemetry/opentelemetry-java/issues/3326

### CODE REVIEW

File: sdk-extensions/autoconfigure/build.gradle.kts [hunk 1]
```diff
@@ -117,22 +122,30 @@ tasks {
...
+    environment("OTEL_METRICS_EXPORTER", "none")
   }
 
+  val testResourceNotInitialized by existing(Test::class)
+
   val check by existing {
     dependsOn(
       testConfigError,
       testFullConfig,
+      testInitializeRegistersGlobal,
```
Comment (mateuszrzeszutek) [inline sdk-extensions/autoconfigure/build.gradle.kts:141]: Turns out those tests weren't even added to the `check` task - I've added and fixed them, they were pretty outdated

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetrySdkAutoConfiguration.java [hunk 2]
```diff
@@ -25,25 +25,30 @@
...
-  /** Returns the automatically configured {@link Resource}. */
+  /**
+   * Returns the automatically configured {@link Resource} after one of {@code initialize()} methods
+   * has been called. Prior to that, this method will always return an {@linkplain Resource#empty()
+   * empty resource}.
+   */
   public static Resource getResource() {
-    return RESOURCE;
+    return resource;
   }
```
Comment (mateuszrzeszutek) [inline sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetrySdkAutoConfiguration.java:37]: I really dislike this part of this PR: now that we allow using custom `ConfigProperties` impl there's a clear temporal dependency between `getResource()` and `initialize()` - which I guess sort of makes sense but I still don't like it.

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java [hunk 3]
```diff
@@ -6,131 +6,55 @@
...
    */
   @Nullable
-  public String getString(String name) {
-    return config.get(name);
-  }
+  String getString(String name);
+
+  /**
+   * Returns a boolean-valued configuration property. Uses the same rules as {@link
+   * Boolean#parseBoolean(String)} for handling the values.
```
Comment (jack-berg) [inline sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java:26]: "Uses the same rules as {@link Boolean#parseBoolean(String)} for handling the values." is an implementation detail.

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java [hunk 4]
```diff
@@ -6,131 +6,55 @@
...
+   * @return false if the property has not been configured.
+   */
+  boolean getBoolean(String name);
 
   /**
    * Returns a integer-valued configuration property.
    *
    * @return null if the property has not been configured.
-   * @throws NumberFormatException if the property is not a valid integer.
+   * @throws ConfigurationException if the property is not a valid integer.
```
Comment (jack-berg) [inline sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java:36]: I'm thinking we should be consistent and document all methods of the interface as having the potential to throw `ConfigurationException`.

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java [hunk 5]
```diff
@@ -6,131 +6,55 @@
...
-  }
+  String getString(String name);
+
+  /**
+   * Returns a boolean-valued configuration property. Uses the same rules as {@link
+   * Boolean#parseBoolean(String)} for handling the values.
+   *
+   * @return false if the property has not been configured.
+   */
+  boolean getBoolean(String name);
```
Comment (jack-berg) [inline sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java:30]: Why does `getBoolean` return a primitive instead of `Boolean`? The behavior is inconsistent with the other methods that return null if the property is unset.

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java [hunk 6]
```diff
@@ -154,117 +78,23 @@ public Double getDouble(String name) {
...
-      char c = rawValue.charAt(lastDigitIndex);
-      if (Character.isDigit(c)) {
-        break;
-      }
-      lastDigitIndex -= 1;
-    }
-    // Pull everything after the last digit.
-    return rawValue.substring(lastDigitIndex + 1);
-  }
+  Map<String, String> getCommaSeparatedMap(String name);
```
Comment (jack-berg) [inline sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java:99]: Don't like that `getCommaSeparatedValues` and `getCommaSeparatedMap` return empty list / map instead of null when not set. I know there's code that depends on that behavior, but if we're making a breaking change by changing this class to an interface, now might be a good time to ensure consistency.

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetrySdkAutoConfiguration.java (hunk 2 above)
Comment (jack-berg) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetrySdkAutoConfiguration.java:37]: This is only the case because of the `getResource()` method. Get rid of this method and the `initialize` methods can operate without any side affects in `OpenTelemetrySdkAutoconfiguration`.

To me, `getResource` feels awkward here. Feels like it could live on `OpenTelemetrySdk` or `OpenTelemetry` instead.

**Review by jack-berg — COMMENTED** [review]
Had a few thoughts on this.

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java (hunk 5 above)
Comment (mateuszrzeszutek) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java:30]: I noticed it too, wasn't sure whether to change it in this PR - but you're right, we're making a breaking change so we might as well fix that one along with it.

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetrySdkAutoConfiguration.java (hunk 2 above)
Comment (mateuszrzeszutek) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetrySdkAutoConfiguration.java:37]: We can't put it into `OpenTelemetry` because the resource is a part of SDK and not API, but `OpenTelemetrySdk` sounds good.
@anuraaga @jkwatson was there any reason for not adding it to the `OpenTelemetrySdk`?

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java (hunk 6 above)
Comment (mateuszrzeszutek) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java:99]: Hmm, I think we're probably returning nulls for Integers/Longs/etc because there's no "natural null object" for those - with a list or a map you can always return an empty one and avoid unneeded null checks.

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java (hunk 3 above)
Comment (jkwatson) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java:26]: I don't agree. I think this is just shorthand for "read those other javadocs to see how this is parsed". The other option would be to copy the contents of that javadoc over here.

Comment (jkwatson) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java:26]: Oh, you're saying this probably doesn't belong on the interface javadoc. Hmm. Perhaps changing it to "Implementations should use the same rules as ..." would be the right thing to do here.

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java (hunk 6 above)
Comment (jkwatson) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java:99]: I don't think we should return nulls. That's a bad API to expose, as all callers will need to do null checking. We should return the empty lists/maps as we do today, IMO.

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetrySdkAutoConfiguration.java (hunk 2 above)
Comment (jkwatson) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetrySdkAutoConfiguration.java:37]: There is a need to be able to get a handle to the Resource that would be autoconfigured, then modify that and use it to build an SDK. So, it would be very cumbersome to have to build an SDK, then get the resource, modify it and rebuild a new SDK.

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java (hunk 6 above)
Comment (jack-berg) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java:99]: I see where you're coming from, but having responses be nullable allows callers to differentiate between "empty" and "not set". 

Don't feel strongly about this.

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetrySdkAutoConfiguration.java (hunk 2 above)
Comment (jkwatson) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetrySdkAutoConfiguration.java:37]: But, I agree that this API is not good, as-is. Perhaps we should split out the Resource autoconfiguration into it's own class. Something like `OpenTelemetryResource.initialize()`

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java (hunk 4 above)
Comment (mateuszrzeszutek) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java:36]: Hmm, even `getString()` or `getCommaSeparatedValues()` which cannot fail? Also, for `getBoolean()` that'd conflict with the `Implementations should use the same rules as Boolean.parseBoolean()` comment.

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetrySdkAutoConfiguration.java (hunk 2 above)
Comment (mateuszrzeszutek) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetrySdkAutoConfiguration.java]: Done, sort of - I've added an `OpenTelemetryResourceAutoConfiguration` class that borrows a few patterns from `GlobalOpenTelemetry` and holds the global auto-configured resource.

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetrySdkAutoConfiguration.java [hunk 7]
```diff
@@ -25,25 +22,20 @@
...
-  /** Returns the automatically configured {@link Resource}. */
-  public static Resource getResource() {
-    return RESOURCE;
-  }
-
   /**
-   * Returns an {@link OpenTelemetrySdk} automatically initialized through recognized system
-   * properties and environment variables.
+   * Returns an {@link OpenTelemetrySdk} automatically initialized through recognized properties
+   * contained in passed {@code config}.
```
Comment (jkwatson) [inline sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetrySdkAutoConfiguration.java:27]:
```suggestion
   * contained in the {@code config} parameter.
```

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetryResourceAutoConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+import io.opentelemetry.sdk.resources.Resource;
+import java.util.HashSet;
+import java.util.ServiceLoader;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/** Auto-configuration for the OpenTelemetry {@link Resource}. */
+public final class OpenTelemetryResourceAutoConfiguration {
+
+  @Nullable private static volatile Resource resource = null;
```
Comment (jkwatson) [inline sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetryResourceAutoConfiguration.java:19]: I wonder...how important is it that we keep a copy of this, making a global of the configured Resource? What would be the downside if we just always re-initialized it?

File: sdk-extensions/autoconfigure/src/testFullConfig/java/io/opentelemetry/sdk/autoconfigure/ResourceTest.java [hunk 9]
```diff
@@ -15,7 +16,10 @@
 
   @Test
   void resource() {
-    Attributes attributes = OpenTelemetrySdkAutoConfiguration.getResource().getAttributes();
+    // make sure that the SDK is autoconfigured first regardless of test execution order
+    GlobalOpenTelemetry.get();
```
Comment (jkwatson) [inline sdk-extensions/autoconfigure/src/testFullConfig/java/io/opentelemetry/sdk/autoconfigure/ResourceTest.java:20]: why is this necessary when you're only testing the Resource?

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetryResourceAutoConfiguration.java (hunk 8 above)
Comment (mateuszrzeszutek) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetryResourceAutoConfiguration.java:19]: Hmm, there'd be a risk of using different `ConfigProperties` for each call. For example, initializing the SDK with a custom `ConfigProperties` implementation and then calling `getResource()` (that uses default config) could result in different resources. 
So, if somebody used the javaagent's `ConfigPropertySource` SPI to set some resource attributes they'd always be visible in the auto-configured SDK, but if you called `OpenTelemetryResource...#getResource()` without passing the  same `ConfigProperties` you'd get a different resource (constructed just from system properties and env vars).

I don't know it it's a problem though, if we document that behavior it should probably be fine. In case of the instrumentation javaagent the config's globally available, so it'd be fine to just ensure we pass it every time we call `getResource()`.

Comment (jkwatson) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetryResourceAutoConfiguration.java:19]: I think if we never try to claim that there's a single global Resource, (maybe "initialize..." isn't the right name for the method), that this would be fine. It's just a factory for Resource instances based on config.

@anuraaga any thoughts on this?

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java (hunk 4 above)
Comment (jack-berg) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java:37]: If someone comes along later with an implementation that isn't based on parsing string values like in `DefaultConfigProperties` then javadoc that implies string parsing will seem strange. A semi-realistic example would be if the config source was YAML or JSON with support for different typed values. It would be appropriate for any of the methods to throw if there was a type mismatch.

Comment (mateuszrzeszutek) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ConfigProperties.java:37]: That's a great point -- I'll add this. Thanks!

File: sdk-extensions/autoconfigure/src/testFullConfig/java/io/opentelemetry/sdk/autoconfigure/ResourceTest.java (hunk 9 above)
Comment (mateuszrzeszutek) [reply sdk-extensions/autoconfigure/src/testFullConfig/java/io/opentelemetry/sdk/autoconfigure/ResourceTest.java]: No longer applicable, removed.

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetryResourceAutoConfiguration.java (hunk 8 above)
Comment (mateuszrzeszutek) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetryResourceAutoConfiguration.java]: Done - PTAL.

Comment (jkwatson) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetryResourceAutoConfiguration.java]: I think this looks good. Would like @anuraaga to chime in before merging.

**Review by jkwatson — APPROVED** [review]
Thanks!

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetrySdkAutoConfiguration.java [hunk 10]
```diff
@@ -61,6 +52,17 @@ public static OpenTelemetrySdk initialize(boolean setResultAsGlobal) {
...
   }
 
+  /**
+   * Returns an {@link OpenTelemetrySdk} automatically initialized through recognized system
+   * properties and environment variables.
+   *
+   * @param setResultAsGlobal Whether to automatically set the configured SDK as the {@link
+   *     io.opentelemetry.api.GlobalOpenTelemetry} instance.
+   */
+  public static OpenTelemetrySdk initialize(boolean setResultAsGlobal) {
```
Comment (anuraaga) [inline sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetrySdkAutoConfiguration.java:62]: Nit we generally order less argument methods over more argument ones

**Review by anuraaga — COMMENTED** [review]
Sorry for coming to this late, last holidays for a while.

It seems like this approach will require a lot of duplication - it's still OK I think but was there a reason to not use the `pass in a Map` approach?

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/DefaultConfigProperties.java [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+/**
+ * Properties to be used for auto-configuration of the OpenTelemetry SDK components. These
+ * properties will be a combination of system properties and environment variables. The properties
+ * for both of these will be normalized to be all lower case, and underscores will be replaced with
+ * periods.
+ */
+final class DefaultConfigProperties implements ConfigProperties {
```
[the remark this answers is not in the document]
Comment (mateuszrzeszutek) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/DefaultConfigProperties.java:27]: Most of it is already there - the javaagent `Config` already parses maps, lists, booleans. The only thing missing is the duration parsing.
So, we won't be copying a lot of new code into the javaagent - but `Config` will end up having the same logic as this class, so there'll be duplication in the end.

**Review by anuraaga — APPROVED** [review]
This change looks good since it unblocks separating out SPI and the java agent from providing the custom config by duplicating the logic. I think we can followup to see if it's worth deduping that with a simpler mechanism.

**Comment by jkwatson** [comment]
I'm going to merge this so it gets into 1.5.0. Let's see how it goes with the agent using this, and we can adjust if we think we can make it simpler.
